File: lib/features/map/domain/usecases/search_stations.dart
import 'package:fpdart/fpdart.dart';
import 'package:station_reach/features/map/domain/models/station.dart';
import 'package:station_reach/features/map/domain/repositories/map_repository.dart';
import 'package:webfabrik_theme/webfabrik_theme.dart';

/// {@template search_stations}
/// Searches for stations by name.
///
/// Parameters:
/// - query: Query [String] to search for
///
/// Returns:
/// - [List] of [Station]s found
///
/// Failures:
/// {@macro converted_dio_exceptions}
/// {@endtemplate}
class SearchStations {
  /// {@macro search_stations}
  SearchStations({required this.mapRepository});

  final MapRepository mapRepository;

  /// {@macro search_stations}
  Future<Either<Failure, List<Station>>> call({required String query}) async {
    return _searchStations(query: query);
  }

  Future<Either<Failure, List<Station>>> _searchStations({
    required String query,
  }) async {
    return mapRepository.searchStations(query: query);
  }
}

File: src/features/map/domain/usecases/search-stations.ts
import type { Result } from 'neverthrow';

import type { Failure } from '@/core/failures';

import type { Station } from '../models/station';
import type { MapRepository } from '../repositories/map-repository';

export type SearchStations = (
  query: string,
  signal?: AbortSignal
) => Promise<Result<Station[], Failure>>;

/** Searches stations by name. */
export function createSearchStations(
  mapRepository: MapRepository
): SearchStations {
  return (query, signal) => mapRepository.searchStations(query, signal);
}
